Hipótesis:

Las hipótesis que os planteamos son:

- ¿Es Australia el sitio más peligroso y letal para estar relajada en la playa?
- ¿Hay diferencias entre los países en los ataques a hombres y mujeres?
- ¿Cuáles son las edades que más sufren ataques?
- Independientemente de la edad, ¿sufren los hombres más ataques que las mujeres.
- ¿En qué mes ocurren más ataques?
- ¿Cuál es la relación entre la especie y el tipo de ataque (si es fatal o no)?
- ¿Cómo han evolucionado los ataques a lo largo del tiempo?

In [37]:
import pandas as pd
import numpy as np
import re
import random as ra

In [38]:
df = pd.read_csv("datos/ataques_pandas7.csv", index_col=0)



**Pair Programming EDA I**

Objetivos
Los objetivos de hoy son:
1. Haced una exploración inicial del dataframe:

¿Cuál es el número de filas y columnas?

In [39]:
df.head(1)

YEAR     TYPE COUNTRY        AREA                     LOCATION  \
CASE_NUMBER                                                                     
2018.06.25   2018.0  Boating     USA  California  Oceanside, San Diego County   

             ACTIVITY         NAME SEX_ AGE   TIME  ... CASE NUMBER.1  \
CASE_NUMBER                                         ...                 
2018.06.25   Paddling  Julie Wolfe    F  57  18h00  ...           NaN   

            CASE NUMBER.2 ORIGINAL ORDER  UNNAMED:22 UNNAMED: 23 UNNAMED: 0_y  \
CASE_NUMBER                                                                     
2018.06.25            NaN            NaN         NaN         NaN         5472   

                                                        INJURY fecha  \
CASE_NUMBER                                                            
2018.06.25   Lacerations to leg from hooked shark PROVOKED ...   NaN   

            FATALIDAD SEX_UAL  
CASE_NUMBER                    
2018.06.25          N       F  

[1 rows x 28 columns]

In [40]:
df.shape

(1672, 28)

El df tiene 1672 filas y 28 columnas.

¿Cuál es el porcentaje de valores nulos en cada columna? Presenta los resultados en un dataframe, donde las columnas sean el nombre de las variables y el porcentaje de nulos.

In [41]:
porcentajes= (df.isnull().sum() * 100) / df.shape[0]

In [42]:
columnas = ['columna', 'porcentaje']

In [43]:
df_null= pd.DataFrame(porcentajes).T

In [44]:
df_null

YEAR  TYPE   COUNTRY      AREA  LOCATION  ACTIVITY      NAME       SEX_  \
0   0.0   0.0  0.598086  2.751196  3.050239  1.854067  0.538278  10.167464   

        AGE      TIME  ...  CASE NUMBER.1  CASE NUMBER.2  ORIGINAL ORDER  \
0  9.210526  9.569378  ...      89.832536      89.832536       89.832536   

   UNNAMED:22  UNNAMED: 23  UNNAMED: 0_y    INJURY      fecha  FATALIDAD  \
0       100.0        100.0           0.0  0.478469  99.461722   0.777512   

   SEX_UAL  
0      0.0  

[1 rows x 28 columns]

¿Qué tipos de datos hay en cada columna?

In [45]:
df.dtypes

YEAR                      float64
TYPE                       object
COUNTRY                    object
AREA                       object
LOCATION                   object
ACTIVITY                   object
NAME                       object
SEX_                       object
AGE                        object
TIME                       object
SPECIES_                   object
PDF                        object
HREF                       object
UNNAMED: 0_x              float64
CASE NUMBER                object
SPECIES                    object
INVESTIGATOR OR SOURCE     object
HREF FORMULA               object
CASE NUMBER.1              object
CASE NUMBER.2              object
ORIGINAL ORDER            float64
UNNAMED:22                float64
UNNAMED: 23               float64
UNNAMED: 0_y                int64
INJURY                     object
fecha                      object
FATALIDAD                  object
SEX_UAL                    object
dtype: object

Extraed la información general del dataframe.

In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1672 entries, 2018.06.25 to nan
Data columns (total 28 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   YEAR                    1672 non-null   float64
 1   TYPE                    1672 non-null   object 
 2   COUNTRY                 1662 non-null   object 
 3   AREA                    1626 non-null   object 
 4   LOCATION                1621 non-null   object 
 5   ACTIVITY                1641 non-null   object 
 6   NAME                    1663 non-null   object 
 7   SEX_                    1502 non-null   object 
 8   AGE                     1518 non-null   object 
 9   TIME                    1512 non-null   object 
 10  SPECIES_                1502 non-null   object 
 11  PDF                     1672 non-null   object 
 12  HREF                    1672 non-null   object 
 13  UNNAMED: 0_x            170 non-null    float64
 14  CASE NUMBER             170 non-null 

2. Cread dos dataframes nuevos, uno que incluya solo las variables categóricas y otro que incluya solo las variables numéricas.

In [47]:
categoricas= df.select_dtypes (include = "object")
categoricas.head()

TYPE    COUNTRY        AREA                      LOCATION  \
CASE_NUMBER                                                                     
2018.06.25       Boating        USA  California   Oceanside, San Diego County   
2018.06.03.a  Unprovoked     BRAZIL  Pernambuco         Piedade Beach, Recife   
2018.05.26.b  Unprovoked        USA     Florida  Cocoa Beach, Brevard  County   
2018.05.24      Provoked  AUSTRALIA  Queensland               Cairns Aquarium   
2018.05.13.a     Invalid    ENGLAND    Cornwall                Off Land's End   

                    ACTIVITY                    NAME SEX_ AGE            TIME  \
CASE_NUMBER                                                                     
2018.06.25          Paddling             Julie Wolfe    F  57           18h00   
2018.06.03.a        Swimming  Jose Ernesto da Silva     M  18  Late afternoon   
2018.05.26.b         Walking               Cody High    M  15           17h00   
2018.05.24    Feeding sharks                    male    M  32         Morning   
2018.05.13.a         Fishing            Max Berryman    M  21           08h15   

                      SPECIES_  ... CASE NUMBER SPECIES   \
CASE_NUMBER                     ...                        
2018.06.25         White shark  ...         NaN      NaN   
2018.06.03.a       Tiger shark  ...         NaN      NaN   
2018.05.26.b    Bull shark, 6'  ...         NaN      NaN   
2018.05.24     Grey reef shark  ...         NaN      NaN   
2018.05.13.a  Invalid incident  ...         NaN      NaN   

             INVESTIGATOR OR SOURCE HREF FORMULA CASE NUMBER.1 CASE NUMBER.2  \
CASE_NUMBER                                                                    
2018.06.25                      NaN          NaN           NaN           NaN   
2018.06.03.a                    NaN          NaN           NaN           NaN   
2018.05.26.b                    NaN          NaN           NaN           NaN   
2018.05.24                      NaN          NaN           NaN           NaN   
2018.05.13.a                    NaN          NaN           NaN           NaN   

                                                         INJURY fecha  \
CASE_NUMBER                                                             
2018.06.25    Lacerations to leg from hooked shark PROVOKED ...   NaN   
2018.06.03.a  Lacerations to leg from hooked shark PROVOKED ...   NaN   
2018.05.26.b  Lacerations to leg from hooked shark PROVOKED ...   NaN   
2018.05.24    Lacerations to leg from hooked shark PROVOKED ...   NaN   
2018.05.13.a  Lacerations to leg from hooked shark PROVOKED ...   NaN   

             FATALIDAD SEX_UAL  
CASE_NUMBER                     
2018.06.25           N       F  
2018.06.03.a         N       M  
2018.05.26.b         N       M  
2018.05.24           N       M  
2018.05.13.a         N       M  

[5 rows x 22 columns]

In [48]:
numericas = df.select_dtypes(include = "float64")
numericas.head()

YEAR  UNNAMED: 0_x  ORIGINAL ORDER  UNNAMED:22  UNNAMED: 23
CASE_NUMBER                                                                
2018.06.25    2018.0           NaN             NaN         NaN          NaN
2018.06.03.a  2018.0           NaN             NaN         NaN          NaN
2018.05.26.b  2018.0           NaN             NaN         NaN          NaN
2018.05.24    2018.0           NaN             NaN         NaN          NaN
2018.05.13.a  2018.0           NaN             NaN         NaN          NaN

¿Cuáles son los principales estádisticos de los dataframes?

In [49]:
numericas.describe()

YEAR  UNNAMED: 0_x  ORIGINAL ORDER  UNNAMED:22  UNNAMED: 23
count  1672.000000    170.000000      170.000000         0.0          0.0
mean   1832.688397   6216.500000       86.500000         NaN          NaN
std     528.969901     49.218899       49.218899         NaN          NaN
min       0.000000   6132.000000        2.000000         NaN          NaN
25%    1964.000000   6174.250000       44.250000         NaN          NaN
50%    1997.000000   6216.500000       86.500000         NaN          NaN
75%    2008.000000   6258.750000      128.750000         NaN          NaN
max    2018.000000   6301.000000      171.000000         NaN          NaN

In [55]:
categoricas.describe()

TYPE COUNTRY     AREA                          LOCATION  \
count         1672    1662     1626                              1621   
unique           6      95      236                              1310   
top     Unprovoked     USA  Florida  New Smyrna Beach, Volusia County   
freq          1432     741      346                                73   

       ACTIVITY  NAME  SEX_   AGE   TIME     SPECIES_  ... CASE NUMBER  \
count      1641  1663  1502  1518   1512         1502  ...         170   
unique      404  1536     3    94    248          834  ...         170   
top     Surfing  male     M    19  11h00  White shark  ...  1800.00.00   
freq        426    97  1294    70     74           74  ...           1   

          SPECIES  INVESTIGATOR OR SOURCE  \
count           44                    169   
unique          35                    142   
top     Bull shark         C. Moore, GSAF   
freq             4                      9   

                                             HREF FORMULA CASE NUMBER.1  \
count                                                 170           170   
unique                                                170           170   
top     http://sharkattackfile.net/spreadsheets/pdf_di...    1800.00.00   
freq                                                    1             1   

       CASE NUMBER.2                                             INJURY fecha  \
count            170                                               1664     9   
unique           170                                                 99     6   
top       1800.00.00  Lacerations to leg from hooked shark PROVOKED ...   Jul   
freq               1                                               1502     2   

       FATALIDAD SEX_UAL  
count       1659    1672  
unique         2       2  
top            N       M  
freq        1573    1378  

[4 rows x 22 columns]

Extraed los valores únicos su frecuencia.

BONUS presenta los resultados en un dataframe.

In [51]:
for col in categoricas.columns:
    print(f"En la columna {col} tenemos los siguientes valores unicos")
    print(categoricas[col].unique())
    print("\n", "------------------------------------------------------------", "\n")

En la columna TYPE tenemos los siguientes valores unicos
['Boating' 'Unprovoked' 'Provoked' 'Invalid' 'Sea Disaster' 'Boat']

 ------------------------------------------------------------ 

En la columna COUNTRY tenemos los siguientes valores unicos
['USA' 'BRAZIL' 'AUSTRALIA' 'ENGLAND' 'COSTA RICA' 'MALDIVES'
 'SOUTH AFRICA' 'CUBA' 'SPAIN' 'BAHAMAS' 'INDONESIA' 'EGYPT' 'NEW ZEALAND'
 'FRENCH POLYNESIA' 'NEW CALEDONIA' 'ECUADOR' 'THAILAND' 'REUNION'
 'MEXICO' 'JAMAICA' 'ITALY' 'PAPUA NEW GUINEA' 'RUSSIA' 'SEYCHELLES'
 'VIETNAM' 'GUAM' 'MOZAMBIQUE' 'SCOTLAND' 'UNITED KINGDOM' 'TAIWAN'
 'CROATIA' 'FIJI' 'TONGA' 'ST. MAARTIN' 'GRAND CAYMAN' 'VANUATU'
 'VENEZUELA' 'TURKS & CAICOS' 'OKINAWA' 'MARSHALL ISLANDS' 'MAURITIUS'
 'HONG KONG' 'JAPAN' 'SOUTH KOREA' 'ISRAEL' 'MICRONESIA' 'COLUMBIA'
 'NEW BRITAIN' 'PALAU' 'SOLOMON ISLANDS' 'FEDERATED STATES OF MICRONESIA'
 'GREECE' 'BRITISH WEST INDIES' 'NORTH PACIFIC OCEAN' 'BERMUDA'
 'PHILIPPINES' 'AMERICAN SAMOA' 'PACIFIC OCEAN' 'SENEGAL' 'ARGENTIN

In [52]:
for col in categoricas.columns:
    print(f"En la columna {col} tenemos los siguientes valores unicos")
    print(pd.DataFrame(categoricas[col].value_counts()))
    print("\n", "------------------------------------------------------------", "\n")

En la columna TYPE tenemos los siguientes valores unicos
              TYPE
Unprovoked    1432
Provoked       115
Invalid         92
Boating         17
Sea Disaster    12
Boat             4

 ------------------------------------------------------------ 

En la columna COUNTRY tenemos los siguientes valores unicos
                    COUNTRY
USA                     741
AUSTRALIA               343
SOUTH AFRICA            234
BAHAMAS                  29
NEW ZEALAND              21
...                     ...
GREENLAND                 1
ST. MAARTIN               1
UNITED KINGDOM            1
SWEDEN                    1
CEYLON (SRI LANKA)        1

[95 rows x 1 columns]

 ------------------------------------------------------------ 

En la columna AREA tenemos los siguientes valores unicos
                                                    AREA
Florida                                              346
California                                           137
New South Wales                  

In [53]:
for col in categoricas.columns:
    pd.DataFrame(categoricas[col].value_counts())


In [54]:
for col in categoricas.columns:
      print(df[col].unique())
      nuevo= pd.DataFrame( df[col], df[col].unique())

['Boating' 'Unprovoked' 'Provoked' 'Invalid' 'Sea Disaster' 'Boat']


C:\Users\marta\AppData\Local\Temp\ipykernel_25532\1308254860.py:3: FutureWarning: reindexing with a non-unique Index is deprecated and will raise in a future version.
  nuevo= pd.DataFrame( df[col], df[col].unique())


ValueError: cannot reindex on an axis with duplicate labels

In [ ]:
for col in numericas.columns:
    print(f"En la columna {col} tenemos los siguientes valores unicos")
    print(numericas[col].unique())
    print("\n", "------------------------------------------------------------", "\n")

En la columna YEAR tenemos los siguientes valores unicos
[2018. 2017. 2016. 2015. 2014. 2013. 2012. 2011. 2010. 2009. 2008. 2007.
 2006. 2005. 2004. 2003. 2002. 2001. 2000. 1999. 1998. 1997. 1996. 1995.
 1994. 1993. 1992. 1991. 1990. 1989. 1988. 1987. 1986. 1985. 1984. 1983.
 1982. 1981. 1980. 1979. 1978. 1977. 1976. 1975. 1974. 1973. 1972. 1971.
 1970. 1969. 1968. 1967. 1966. 1965. 1964. 1963. 1962. 1961. 1960. 1959.
 1958. 1957. 1956. 1955. 1954. 1953. 1952. 1951. 1950. 1949. 1948. 1947.
 1946. 1945. 1944. 1943. 1942. 1941. 1940. 1939. 1938. 1937. 1936. 1935.
 1934. 1932. 1931. 1930. 1929. 1928. 1927. 1926. 1925. 1924. 1922. 1919.
 1916. 1912. 1906. 1905. 1901. 1894. 1892. 1880. 1868. 1800. 1797. 1792.
 1791. 1788. 1787. 1786. 1785. 1784. 1783. 1780. 1779. 1776. 1771. 1767.
 1764. 1758. 1753. 1751. 1749. 1755. 1748. 1742. 1738. 1733. 1723. 1721.
 1703. 1700. 1642. 1638. 1637. 1617. 1595. 1580. 1555. 1554. 1543.  500.
   77.    5.    0.]

 ---------------------------------------------

In [ ]:
for col in numericas.columns:
    print(f"En la columna {col} tenemos los siguientes valores unicos")
    print(pd.DataFrame(numericas[col].value_counts()))
    print("\n", "------------------------------------------------------------", "\n")

En la columna YEAR tenemos los siguientes valores unicos
        YEAR
0.0      125
2015.0    68
2007.0    57
2012.0    54
2002.0    53
...      ...
1785.0     1
1783.0     1
1780.0     1
1779.0     1
1784.0     1

[147 rows x 1 columns]

 ------------------------------------------------------------ 

En la columna UNNAMED: 0_x tenemos los siguientes valores unicos
        UNNAMED: 0_x
6132.0             1
6248.0             1
6240.0             1
6241.0             1
6242.0             1
...              ...
6190.0             1
6191.0             1
6192.0             1
6193.0             1
6301.0             1

[170 rows x 1 columns]

 ------------------------------------------------------------ 

En la columna ORIGINAL ORDER tenemos los siguientes valores unicos
       ORIGINAL ORDER
171.0               1
55.0                1
63.0                1
62.0                1
61.0                1
...               ...
113.0               1
112.0               1
111.0               1
110.0

Conclusiones:


¿De qué nos sirven estos números y dataframe? Extraed conclusiones de ellos.


Por ejemplo:
- ¿Hay alguna columna que no tenga sentido su tipo de datos?
- ¿Hay alguna columna que no tenga sentido su nombre?
- ¿Hay alguna columna que no tenga sentido su contenido?
- ¿Hay alguna columna que tenga demasiadas categorías o valores que nos pueda hacer difícil trabajar con ella?

In [ ]:

# En la columna NAME hay bastantes datos que no se corresponden a esta categoria y bnos dicen el género de la persona afectada y hasta la categoría de edad de 
# alguno de ellos (child).

# En la columna SEX hay un dato llamado lli que no se corresponde a ningun género.

# En la columna TIME las categorías no están ordenadas. Hay algunos datos que estan por franjas horarias que abarcan un número de horas mientras que otros estan
# en horas concretas.

# La columna SPECIES no aporta el nombre de la especie como tal si no un abánico en el que hay desde descripciones ambiguas hasta especificaciones de tamaños y pesos.
